### PR TITLE
chore: add agentapi tests

### PR DESCRIPTION
**Pull request by deansheather**
~~The tests added in this PR can be reviewed now as I probably won't be changing them (other than review comments). I'll be adding more tests for the files TODOd below in upcoming commits and I'll re-request review.~~

In most cases the new tests have more cases/coverage than the old ones which is nice. Some of them are quite verbose though, as I tried to maximize usage of fields on types and enums (to get conversion coverage).

## TODOs:

- [x] `metadata.go`
- [x] `stats.go`
- [x] `tailnet.go`

### CODE REVIEW

**Comment by deansheather** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #11269** <a href="https://app.graphite.dev/github/pr/coder/coder/11269?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈
    * **PR #11280** <a href="https://app.graphite.dev/github/pr/coder/coder/11280?utm_source=stack-comment-icon" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/coder/coder/11269?utm_source=stack-comment).

**Review by spikecurtis — COMMENTED** [review]
I didn't finish this review, but I'm sending the comments I have since you're in a different timezone

File: coderd/agentapi/metadata_test.go [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+			require.NoError(t, json.Unmarshal(message, &gotEvent))
+		})
+		require.NoError(t, err)
+		defer cancel()
+
+		resp, err := api.BatchUpdateMetadata(context.Background(), req)
+		require.NoError(t, err)
+		require.Equal(t, &agentproto.BatchUpdateMetadataResponse{}, resp)
+
+		require.Equal(t, int64(1), atomic.LoadInt64(&eventCount))
```
Comment (spikecurtis) [inline coderd/agentapi/metadata_test.go:101]: Pubsub calls the subscriber on a goroutine, which means that this test can race.

You could create a fake pubsub that only implements `Publish()` (the rest can panic), and stores the results internally, then assert on the state of the fake pubsub.

An even better change would be to break up the Pubsub interface like

```
type Publisher interface {
    Publish(event string, message []byte) error
}

type Subscriber interface {
   Subscribe(event string, listener Listener) (cancel func(), err error)
   SubscribeWithErr(event string, listener ListenerWithErr) (cancel func(), err error)
}

type Pubsub interface {
    Publisher
    Subscriber
    io.Closer
}
```

then the `MetadataAPI` can take a `Publisher` and you only need to implement `Publish()` on your fake.

File: coderd/agentapi/stats_test.go [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+		api := agentapi.StatsAPI{
+			AgentFn: func(context.Context) (database.WorkspaceAgent, error) {
+				return agent, nil
+			},
+			Database:                  dbM,
+			StatsBatcher:              batcher,
+			TemplateScheduleStore:     templateScheduleStorePtr(templateScheduleStore),
+			AgentStatsRefreshInterval: 10 * time.Second,
+			UpdateAgentMetricsFn: func(ctx context.Context, labels prometheusmetrics.AgentMetricLabels, metrics []*agentproto.Stats_Metric) {
+				atomic.AddInt64(&updateAgentMetricsFnCalled, 1)
```
Comment (spikecurtis) [inline coderd/agentapi/stats_test.go:132]: you never check that this was incremented

File: coderd/agentapi/stats_test.go [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+		api := agentapi.StatsAPI{
+			AgentFn: func(context.Context) (database.WorkspaceAgent, error) {
+				return agent, nil
+			},
+			Database:                  dbM,
+			StatsBatcher:              batcher,
+			TemplateScheduleStore:     templateScheduleStorePtr(templateScheduleStore),
+			AgentStatsRefreshInterval: 15 * time.Second,
+			UpdateAgentMetricsFn: func(ctx context.Context, labels prometheusmetrics.AgentMetricLabels, metrics []*agentproto.Stats_Metric) {
+				atomic.AddInt64(&updateAgentMetricsFnCalled, 1)
```
Comment (spikecurtis) [inline coderd/agentapi/stats_test.go:327]: you never check this was incremented

File: coderd/agentapi/stats_test.go [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+			String: "CRON_TZ=Australia/Sydney 0 8 * * *",
+			Valid:  true,
+		}
+
+		// Use a custom time for now which would trigger the autostart aware
+		// bump.
+		now, err := time.Parse("2006-01-02 15:04:05 -0700 MST", "2023-12-19 07:30:00 +1100 AEDT")
+		require.NoError(t, err)
+		now = dbtime.Time(now)
+		nextAutostart := now.Add(30 * time.Minute).UTC() // always sent to DB as UTC
```
Comment (spikecurtis) [inline coderd/agentapi/stats_test.go:288]: to me, this smells like we need to refactor.  The AgentAPI should not be in the business of computing autostart times.  There should be a scheduler component that gets an activity indication from the Agent API and it computes and decides how to update the autostart / autostop schedule.

Another one for some other PR...

**Review by spikecurtis — APPROVED** [review]

**Pull request merged**